### PR TITLE
[Testing/Tooling] Refactor test invoke tasks looping over flavors and modules

### PR DESCRIPTION
**Pull request by Pythyu**
### What does this PR do?

Refactor of 4 test functions that are similar, as they perform some kind of action on all flavors and modules passed in parameters.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by Pythyu** [comment]
Forgot an argument of `test_core`. Fixing it right now

File: tasks/test.py [hunk 1]
```diff
@@ -292,14 +305,7 @@ def test_flavor(
...
 
-    for module in modules:
-        module_test_result = ModuleTestResult(path=module.full_path())
-
-        print(f"----- Module '{module.full_path()}'")
-        if not module.condition():
-            print("----- Skipped")
-            continue
-
+    def command(modules_test_results, module, module_test_result):
```
Comment (amenasria) [inline tasks/test.py:308]:
```suggestion
    def command(test_results, module, module_result):
```
The old variables name were not really specific. I think we could replace:
- `modules_test_results` with `test_results`
- `module_test_result` with `module_result`

WDYT? If you have a better idea on how to differentiate them, feel free to share it !

File: tasks/test.py [hunk 2]
```diff
@@ -129,21 +129,38 @@ def install_tools(ctx):
 # wrapper functions that craft the command to run and process the results and errors.
```
Comment (amenasria) [inline tasks/test.py:129]:
```suggestion

```
and all the above comment since it's solved by your PR !

**Review by amenasria — CHANGES_REQUESTED** [review]

File: tasks/test.py [hunk 3]
```diff
@@ -129,21 +129,38 @@ def install_tools(ctx):
...
+    exec_function,
+    skip_module_class=False,
 ):
-    """
-    Runs linters for given flavor, build tags, and modules.
-    """
-    modules_lint_results = []
-    print(f"--- Flavor {flavor.name}: golangci_lint")
+    modules_results = []
+    print(f"--- Flavor {flavor.name}: ", flavor_print_string)
```
[the remark this answers is not in the document]
Comment (Pythyu) [reply tasks/test.py]: Applied in [d99e2ad](https://github.com/DataDog/datadog-agent/pull/15957/commits/d99e2ad6900a41769652c2c2576935084bc0910a)

File: tasks/test.py [hunk 4]
```diff
@@ -129,21 +129,38 @@ def install_tools(ctx):
...
     for module in modules:
-        module_lint_result = ModuleLintResult(path=module.full_path())
+        if not skip_module_class:
+            module_result = module_class(path=module.full_path())
+        else:
+            module_result = None
```
[the remark this answers is not in the document]
Comment (Pythyu) [reply tasks/test.py]: Applied in [d99e2ad](https://github.com/DataDog/datadog-agent/pull/15957/commits/d99e2ad6900a41769652c2c2576935084bc0910a)

File: tasks/test.py [hunk 5]
```diff
@@ -129,21 +129,38 @@ def install_tools(ctx):
...
-def lint_flavor(
-    ctx, modules: List[GoModule], flavor: AgentFlavor, build_tags: List[str], arch: str, rtloader_root: bool
+def test_core(
+    modules: List[GoModule],
+    flavor: AgentFlavor,
+    module_class: GoModule,
+    flavor_print_string: str,
+    exec_function,
+    skip_module_class=False,
 ):
```
[the remark this answers is not in the document]
Comment (Pythyu) [reply tasks/test.py]: Added in [d99e2ad](https://github.com/DataDog/datadog-agent/pull/15957/commits/d99e2ad6900a41769652c2c2576935084bc0910a)

File: tasks/test.py (hunk 1 above)
Comment (Pythyu) [reply tasks/test.py]: I applied for the lint_test as well : 
  - `modules_test_results` with `test_results`
  - `module_test_result` with `module_result`
  - `modules_lint_results` with `lint_results`
  - `module_lint_result` with `module_result`
 [d99e2ad](https://github.com/DataDog/datadog-agent/pull/15957/commits/d99e2ad6900a41769652c2c2576935084bc0910a)

File: tasks/test.py [hunk 6]
```diff
@@ -124,37 +124,51 @@ def install_tools(ctx):
...
-# wrapper functions that craft the command to run and process the results and errors.
-
-
-def lint_flavor(
-    ctx, modules: List[GoModule], flavor: AgentFlavor, build_tags: List[str], arch: str, rtloader_root: bool
+def test_core(
+    modules: List[GoModule],
+    flavor: AgentFlavor,
+    module_class: GoModule,
+    flavor_print_string: str,
```
Comment (amenasria) [inline tasks/test.py:131]:
```suggestion
    operation_name: str,
```

File: tasks/test.py [hunk 7]
```diff
@@ -124,37 +124,51 @@ def install_tools(ctx):
...
+    skip_module_class=False,
 ):
     """
-    Runs linters for given flavor, build tags, and modules.
+    Run the command function on each module of the modules list.
     """
-    modules_lint_results = []
-    print(f"--- Flavor {flavor.name}: golangci_lint")
+    modules_results = []
+    print(f"--- Flavor {flavor.name}: {flavor_print_string}")
```
Comment (amenasria) [inline tasks/test.py:139]:
```suggestion
    print(f"--- Flavor {flavor.name}: {operation_name}")
```

**Review by amenasria — CHANGES_REQUESTED** [review]
Last small nit

File: tasks/test.py (hunk 7 above)
Comment (Pythyu) [reply tasks/test.py]: Fixed in [b4540b7](https://github.com/DataDog/datadog-agent/pull/15957/commits/b4540b7e10da28b6052cd07029be7a14f9bb7213)

File: tasks/test.py (hunk 6 above)
Comment (Pythyu) [reply tasks/test.py]: [b4540b7](https://github.com/DataDog/datadog-agent/pull/15957/commits/b4540b7e10da28b6052cd07029be7a14f9bb7213)

**Review by amenasria — APPROVED** [review]
LGTM ! 🎉

File: tasks/test.py [hunk 8]
```diff
@@ -124,37 +124,51 @@ def install_tools(ctx):
...
+
+
+def lint_flavor(
+    ctx, modules: List[GoModule], flavor: AgentFlavor, build_tags: List[str], arch: str, rtloader_root: bool
+):
+    """
+    Runs linters for given flavor, build tags, and modules.
+    """
+
+    def command(lint_results, module, module_result):
```
Comment (KSerrania) [inline tasks/test.py:160]: In this method, the same `lint_results` variable is used for three distinct things, which makes its content unexpected:
- it is one of the parameters of the `command` function, which should append a new `ModuleLintResult` to it 
- however, `lint_results` is then redefined (line 162) as the output of `run_golangci_lint`, ie. a new list
- therefore, at this point, `lint_results` doesn't point to the list that was passed to the function anymore, and therefore the `lint_results.append` on line 169 doesn't really do anything.

The end result is that `lint_flavor` always returns an empty list, meaning that we are not catching linter errors.

**Pull request merged**